File: 90.子集-ii.py
#
# @lc app=leetcode.cn id=90 lang=python3
#
# [90] 子集 II
#

# @lc code=start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        # 划重点：包含重复元素
        res = []
        nums.sort() # diff
        def backtrack(start, tmp):
            res.append(tmp.copy())
            for i in range(start, len(nums)):
                if i > start and nums[i-1] == nums[i]: # diff 保证在同一层不重复
                    continue
                tmp.append(nums[i])
                backtrack(i+1, tmp)
                tmp.pop()
        backtrack(0, [])
        return res
# @lc code=end

File: 90.子集-ii.java
/*
 * @lc app=leetcode.cn id=90 lang=java
 *
 * [90] 子集 II
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

class Solution {
    List<List<Integer>> res = new ArrayList<>();
    //集合和序列有区别
    //4,4,2,4,4 考虑 442和244
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        backtrack(0, nums, new ArrayList<>());
        return res;
    }

    public void backtrack(int start, int[] nums, List<Integer> tmp){
        res.add(new ArrayList<>(tmp));
        for(int i=start;i<nums.length;i++){
            if(i>start&&nums[i]==nums[i-1])continue;
            tmp.add(nums[i]);
            backtrack(i+1, nums, tmp);
            tmp.remove(tmp.size()-1);
        }
    }
}
// @lc code=end

